packages/context/src/promise-helper.ts

```diff
@@ -0,0 +1,101 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {BoundValue, ValueOrPromise} from './binding';
+
+/**
+ * Check whether a value is a Promise-like instance.
+ * Recognizes both native promises and third-party promise libraries.
+ *
+ * @param value The value to check.
+ */
+export function isPromise<T>(
+  value: T | PromiseLike<T>,
+): value is PromiseLike<T> {
+  if (!value) return false;
+  if (typeof value !== 'object' && typeof value !== 'function') return false;
+  return typeof (value as PromiseLike<T>).then === 'function';
+}
+
+/**
+ * A type for a value of `T`, or a promise of `T | RejectionError`.
+ */
+export type ValueOrPromiseWithError<T> = T | Promise<T | RejectionError>;
+
+/**
+ * A special Error to wrap a rejected promise or another Error. It allows us
+ * to propagate errors gracefully through the promise/async/await chain without
+ * UnhandledPromiseRejectionWarning.
```

Comment:
Could you please explain why we would trigger `UnhandledPromiseRejectionWarning` if we were reporting this kind of errors directly via `throw`/`Promise.reject` and then modifying the code which is checking for `RejectionError` to use `catch` or `Promise.prototype.catch` instead of checking the type of the resolved value?

Reply:
You can ignore this comment now, see https://github.com/strongloop/loopback-next/pull/671#discussion_r160131295